src/Conversion/ONNXToKrnl/Math/Elementwise.cpp

```diff
@@ -1385,6 +1388,122 @@ static LogicalResult getVariadicSimdCodeFullyFlattened(
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// Utilities for Op fusion at lowering
+//===----------------------------------------------------------------------===//
+
+// Function pointer type for emitScalarOpFor<T>
+typedef Value (*EmitScalarFunc)(ConversionPatternRewriter &rewriter,
+    Location loc, Operation *op, Type elementType,
+    ArrayRef<Value> scalarOperands);
+
+// Variadic template to iterate all the fusible Ops
+template <typename T>
+bool enqueueFusedOpImpl(Operation *op, SmallVector<Operation *, 2> &fusibleOps,
+    SmallVector<EmitScalarFunc, 2> &fuseEmitFunctions) {
+  if (isa<T>(op)) {
+    fusibleOps.emplace_back(op);
+    fuseEmitFunctions.emplace_back(emitScalarOpFor<T>);
+    return true;
+  } else {
+    return false;
+  }
+}
+
+template <typename T = void, class... Ts>
+bool enqueueFusedOp(Operation *op, SmallVector<Operation *, 2> &fusibleOps,
+    SmallVector<EmitScalarFunc, 2> &fuseEmitFunctions);
+
+template <typename T, class... Ts>
+bool enqueueFusedOp(Operation *op, SmallVector<Operation *, 2> &fusibleOps,
+    SmallVector<EmitScalarFunc, 2> &fuseEmitFunctions) {
+  if (enqueueFusedOpImpl<T>(op, fusibleOps, fuseEmitFunctions)) {
+    return true;
+  } else {
+    return enqueueFusedOp<Ts...>(op, fusibleOps, fuseEmitFunctions);
+  }
+}
+
+template <>
+bool enqueueFusedOp(Operation *op, SmallVector<Operation *, 2> &fusibleOps,
+    SmallVector<EmitScalarFunc, 2> &fuseEmitFunctions) {
+  return false;
+}
+
+bool enqueueFusedOp(Operation *op, SmallVector<Operation *, 2> &fusibleOps,
+    SmallVector<EmitScalarFunc, 2> &fuseEmitFunctions) {
+
+  // Notice: Though ClipOp is classified as unary element op in this file,
+  // ClipOp requires one required input and two optional input
+
+  return enqueueFusedOp<ONNXCastOp, ONNXSinhOp, ONNXCoshOp, ONNXSigmoidOp,
+      ONNXHardSigmoidOp, ONNXEluOp, ONNXReluOp, ONNXLeakyReluOp, ONNXPReluOp,
+      ONNXSeluOp, ONNXReciprocalOp, ONNXSoftplusOp, ONNXSoftsignOp, ONNXSignOp,
+      ONNXMaxOp, ONNXMinOp, ONNXNegOp, ONNXLessOp, ONNXLessOrEqualOp,
+      ONNXGreaterOp, ONNXGreaterOrEqualOp, ONNXEqualOp, ONNXNotOp, ONNXModOp,
+      ONNXMeanOp, ONNXRoundOp, ONNXDequantizeLinearOp, ONNXSqrtOp>(
+      op, fusibleOps, fuseEmitFunctions);
+}
+
+// This function starts for elementwise operation, op.
+// A successor op (user) is fusible if it's the only user and an unary
+// elementwise Op. The Op and its EmitScalarOpFor<T> are recorded into
+// the vector.
+static bool findFusibleUnaryOps(Operation *op,
+    SmallVector<Operation *, 2> &fusibleOps,
```

Comment:
I often see the use of `SmallVectorImpl<Operation*>` function argument, instead of `SmallVector<Operation *, 2>`. Same for the other functions in this patch.

Reply:
What's the difference?

Reply:
`SmallVector<Operation *, 2>` requires that users of the function must declare a vector with exact type of `SmallVector<Operation *, 2>` to pass in, while with `SmallVectorImpl<Operation*>`, users can declare a vector of `SmallVector<Operation *, 4>` for example. It's more flexible.

Description of `SmallVectorImpl` in LLVM: https://llvm.org/doxygen/classllvm_1_1SmallVectorImpl.html#details
```
This class consists of common code factored out of the SmallVector class to reduce code duplication based on the SmallVector 'N' template parameter.
```